In [1]:
import sympy as smp

r1, r2 = smp.symbols('r1 r2 ', real=True, positive=True)
theta1, phi1 = smp.symbols('theta_1 phi_1')
theta2, phi2 = smp.symbols('theta_2 phi_2')

Psi = smp.Function('Psi')(r1, theta1, phi1, r2, theta2, phi2)
Psi = smp.exp(-2 * (r1 + r2))

In [2]:
theta = smp.symbols('theta')
absolute_diff = smp.sqrt(r1**2 + r2**2 - 2*r1*r2*smp.cos(theta))

V_eff = -2/r1 - 2/r2 + 1/(absolute_diff)
Laplacian = (1/r1**2) * smp.diff(r1**2 * smp.diff(Psi, r1), r1) + (1/r2**2) * smp.diff(r2**2 * smp.diff(Psi, r2), r2) + (1/(r1**2 * smp.sin(theta1))) * smp.diff(smp.sin(theta1) * smp.diff(Psi, theta1), theta1) + (1/(r2**2 * smp.sin(theta2))) * smp.diff(smp.sin(theta2) * smp.diff(Psi, theta2), theta2) + (1/(r1**2 * smp.sin(theta1)**2)) * smp.diff(smp.diff(Psi, phi1), phi1) + (1/(r2**2 * smp.sin(theta2)**2)) * smp.diff(smp.diff(Psi, phi2), phi2)

In [3]:
x1 = smp.Function(r'x1')(r1, theta1, phi1)
x2 = smp.Function(r'x2')(r2, theta2, phi2)

y1 = smp.Function(r'y1')(r1, theta1, phi1)
y2 = smp.Function(r'y2')(r2, theta2, phi2)

z1 = smp.Function(r'z1')(r1, theta1)
z2 = smp.Function(r'z2')(r2, theta2)

In [4]:
x1 = r1 * smp.sin(theta1) * smp.cos(phi1)
x2 = r2 * smp.sin(theta2) * smp.cos(phi2)

y1 = r1 * smp.sin(theta1) * smp.sin(phi1)
y2 = r2 * smp.sin(theta2) * smp.sin(phi2)

z1 = r1 * smp.cos(theta1)
z2 = r2 * smp.cos(theta2)

In [5]:
# r1 and r2 vector in cartesian coordinates
r1_cartesian = smp.Matrix([x1, y1, z1])
r2_cartesian = smp.Matrix([x2, y2, z2])

# Compute the dot product
dot_product = r1_cartesian.dot(r2_cartesian)

# Compute the magnitude of r1 and r2
magnitude_r1 = smp.sqrt(x1**2 + y1**2 + z1**2)
magnitude_r2 = smp.sqrt(x2**2 + y2**2 + z2**2)

# Compute cos(theta) and sin(theta)
cos_theta = dot_product / (magnitude_r1 * magnitude_r2)

In [6]:
cos_theta.simplify()

sin(theta_1)*sin(theta_2)*cos(phi_1 - phi_2) + cos(theta_1)*cos(theta_2)

In [7]:
V_eff = V_eff.subs(smp.cos(theta), cos_theta)

H_Psi = Laplacian + V_eff * Psi

In [8]:
f = Psi * H_Psi * r1**2 * r2**2 * smp.sin(theta1) * smp.sin(theta2)
f = smp.simplify(f)
f

r1*r2*(8*r1*r2 + r1*r2/sqrt(r1**2 - 2*r1*r2*sin(theta_1)*sin(theta_2)*cos(phi_1 - phi_2) - 2*r1*r2*cos(theta_1)*cos(theta_2) + r2**2) - 6*r1 - 6*r2)*exp(-4*r1 - 4*r2)*sin(theta_1)*sin(theta_2)

In [9]:
g = Psi * Psi * r1**2 * r2**2 * smp.sin(theta1) * smp.sin(theta2)
g = smp.simplify(g)
g

r1**2*r2**2*exp(-4*r1 - 4*r2)*sin(theta_1)*sin(theta_2)

In [10]:
Sampling_func = Psi * Psi
Sampling_func

exp(-4*r1 - 4*r2)

In [11]:
f_f = smp.lambdify([r1, theta1, phi1, r2, theta2, phi2], f)
g_f = smp.lambdify([r1, theta1, phi1, r2, theta2, phi2], g)
Sampling_func_f = smp.lambdify([r1, theta1, phi1, r2, theta2, phi2], Sampling_func)

In [12]:
import numpy as np

r1 = np.linspace(0, 5, 1000)
r2 = np.linspace(0, 5, 1000)

theta1 = np.linspace(0, np.pi, 1000)
theta2 = np.linspace(0, np.pi, 1000)

phi1 = np.linspace(0, 2*np.pi, 1000)
phi2 = np.linspace(0, 2*np.pi, 1000)

R1, R2 = np.meshgrid(r1, r2)
Theta1, Theta2 = np.meshgrid(theta1, theta2)
Phi1, Phi2 = np.meshgrid(phi1, phi2)

In [13]:
f_3d = f_f(R1, Theta1, Phi1, R2, Theta2, Phi2)
g_3d = g_f(R1, Theta1, Phi1, R2, Theta2, Phi2)
Sampling_func_3d = Sampling_func_f(R1, Theta1, Phi1, R2, Theta2, Phi2)

<lambdifygenerated-1>:2: RuntimeWarning: invalid value encountered in sqrt
  return r1*r2*(8*r1*r2 + r1*r2/sqrt(r1**2 - 2*r1*r2*sin(theta_1)*sin(theta_2)*cos(phi_1 - phi_2) - 2*r1*r2*cos(theta_1)*cos(theta_2) + r2**2) - 6*r1 - 6*r2)*exp(-4*r1 - 4*r2)*sin(theta_1)*sin(theta_2)
<lambdifygenerated-1>:2: RuntimeWarning: divide by zero encountered in divide
  return r1*r2*(8*r1*r2 + r1*r2/sqrt(r1**2 - 2*r1*r2*sin(theta_1)*sin(theta_2)*cos(phi_1 - phi_2) - 2*r1*r2*cos(theta_1)*cos(theta_2) + r2**2) - 6*r1 - 6*r2)*exp(-4*r1 - 4*r2)*sin(theta_1)*sin(theta_2)
<lambdifygenerated-1>:2: RuntimeWarning: invalid value encountered in divide
  return r1*r2*(8*r1*r2 + r1*r2/sqrt(r1**2 - 2*r1*r2*sin(theta_1)*sin(theta_2)*cos(phi_1 - phi_2) - 2*r1*r2*cos(theta_1)*cos(theta_2) + r2**2) - 6*r1 - 6*r2)*exp(-4*r1 - 4*r2)*sin(theta_1)*sin(theta_2)


In [14]:
import pandas as pd

# Flatten the arrays
R1_flat = R1.flatten()
R2_flat = R2.flatten()
Theta1_flat = Theta1.flatten()
Theta2_flat = Theta2.flatten()
Phi1_flat = Phi1.flatten()
Phi2_flat = Phi2.flatten()
F_flat = f_3d.flatten()
G_flat = g_3d.flatten()
Sampling_func_flat = Sampling_func_3d.flatten()


# Create the DataFrame
df = pd.DataFrame({'r1': R1_flat, 'theta1': Theta1_flat, 'phi1': Phi1_flat, 'r2': R2_flat, 'theta2': Theta2_flat, 'phi2': Phi2_flat, 'F': F_flat, 'G': G_flat, 'Sampling_func': Sampling_func_flat})

# Display the DataFrame
df

r1    theta1      phi1   r2    theta2      phi2             F  \
0       0.000000  0.000000  0.000000  0.0  0.000000  0.000000           NaN   
1       0.005005  0.003145  0.006289  0.0  0.000000  0.000000 -0.000000e+00   
2       0.010010  0.006289  0.012579  0.0  0.000000  0.000000 -0.000000e+00   
3       0.015015  0.009434  0.018868  0.0  0.000000  0.000000 -0.000000e+00   
4       0.020020  0.012579  0.025158  0.0  0.000000  0.000000 -0.000000e+00   
...          ...       ...       ...  ...       ...       ...           ...   
999995  4.979980  3.129014  6.258027  5.0  3.141593  6.283185  9.131591e-32   
999996  4.984985  3.132158  6.264317  5.0  3.141593  6.283185  8.362173e-32   
999997  4.989990  3.135303  6.270606  5.0  3.141593  6.283185  7.617010e-32   
999998  4.994995  3.138448  6.276896  5.0  3.141593  6.283185  6.895497e-32   
999999  5.000000  3.141593  6.283185  5.0  3.141593  6.283185           inf   

                   G  Sampling_func  
0       0.000000e+00   1.000000e+00  
1       0.000000e+00   9.801790e-01  
2       0.000000e+00   9.607510e-01  
3       0.000000e+00   9.417080e-01  
4       0.000000e+00   9.230424e-01  
...              ...            ...  
999995  4.395801e-33   4.602556e-18  
999996  3.238040e-33   4.511329e-18  
999997  2.120175e-33   4.421910e-18  
999998  1.041166e-33   4.334263e-18  
999999  3.982194e-47   4.248354e-18  

[1000000 rows x 9 columns]

In [15]:
# Remove rows with nan or inf values in 'F11' or 'P1' columns
df_new = df.replace([np.inf, -np.inf], np.nan).dropna(subset=['F','G', 'Sampling_func'])

# Display the cleaned DataFrame
df_new

r1    theta1      phi1   r2    theta2      phi2             F  \
1       0.005005  0.003145  0.006289  0.0  0.000000  0.000000 -0.000000e+00   
2       0.010010  0.006289  0.012579  0.0  0.000000  0.000000 -0.000000e+00   
3       0.015015  0.009434  0.018868  0.0  0.000000  0.000000 -0.000000e+00   
4       0.020020  0.012579  0.025158  0.0  0.000000  0.000000 -0.000000e+00   
5       0.025025  0.015724  0.031447  0.0  0.000000  0.000000 -0.000000e+00   
...          ...       ...       ...  ...       ...       ...           ...   
999994  4.974975  3.125869  6.251738  5.0  3.141593  6.283185  9.925886e-32   
999995  4.979980  3.129014  6.258027  5.0  3.141593  6.283185  9.131591e-32   
999996  4.984985  3.132158  6.264317  5.0  3.141593  6.283185  8.362173e-32   
999997  4.989990  3.135303  6.270606  5.0  3.141593  6.283185  7.617010e-32   
999998  4.994995  3.138448  6.276896  5.0  3.141593  6.283185  6.895497e-32   

                   G  Sampling_func  
1       0.000000e+00   9.801790e-01  
2       0.000000e+00   9.607510e-01  
3       0.000000e+00   9.417080e-01  
4       0.000000e+00   9.230424e-01  
5       0.000000e+00   9.047468e-01  
...              ...            ...  
999994  5.594520e-33   4.695628e-18  
999995  4.395801e-33   4.602556e-18  
999996  3.238040e-33   4.511329e-18  
999997  2.120175e-33   4.421910e-18  
999998  1.041166e-33   4.334263e-18  

[999332 rows x 9 columns]

In [16]:
X_features = df_new.iloc[:, :6].to_numpy()
yf_labels = df_new.iloc[:, 6:8].to_numpy()
yp_labels = df_new.iloc[:, 8:].to_numpy()

print(X_features.shape)
print(yf_labels.shape)
print(yp_labels.shape)

(999332, 6)
(999332, 2)
(999332, 1)


In [17]:
# Define the function
def f(X):
    r1, theta1, phi1, r2, theta2, phi2 = X  # Here X is a vector with 6 components

    result = f_f(r1, theta1, phi1, r2, theta2, phi2) # f_11_f was the lambdified function which converts symbolic expression into numpy arrays

    if 0 <= r1 <= 5 and 0 <= r2 <= 5 and 0 <= theta1 <= np.pi and 0 <= theta2 <= np.pi and 0 <= phi1 <= 2*np.pi and 0 <= phi2 <= 2*np.pi :
        # Check for NaN or inf values
        if np.isnan(result) or np.isinf(result):
          return -np.inf  ## The function will return -infinity if the result is either nan (0/0 or oo/oo) or infinity

        else:
          return result

    else:
        return 0   ## The function will return 0 if any of the r1, r2, theta1, theta2, phi1 and phi2 lies outside of their range

def g(X):
    r1, theta1, phi1, r2, theta2, phi2 = X  # Here X is a vector with 6 components

    result = g_f(r1, theta1, phi1, r2, theta2, phi2) # f_11_f was the lambdified function which converts symbolic expression into numpy arrays

    if 0 <= r1 <= 5 and 0 <= r2 <= 5 and 0 <= theta1 <= np.pi and 0 <= theta2 <= np.pi and 0 <= phi1 <= 2*np.pi and 0 <= phi2 <= 2*np.pi :
        # Check for NaN or inf values
        if np.isnan(result) or np.isinf(result):
          return -np.inf  ## The function will return -infinity if the result is either nan (0/0 or oo/oo) or infinity

        else:
          return result

    else:
        return 0   ## The function will return 0 if any of the r1, r2, theta1, theta2, phi1 and phi2 lies outside of their range

def p(X):
    r1, theta1, phi1, r2, theta2, phi2 = X

    result_p = Sampling_func_f(r1, theta1, phi1, r2, theta2, phi2)

    if 0 <= r1 <= 5 and 0 <= r2 <= 5 and 0 <= theta1 <= np.pi and 0 <= theta2 <= np.pi and 0 <= phi1 <= 2*np.pi and 0 <= phi1 <= 2*np.pi :
        # Check for NaN or inf values
        if np.isnan(result_p) or np.isinf(result_p):
          return -np.inf

        else:
          return result_p

    else:
        return 0

def metropolis_sampling(f,p,initial, num_samples, proposal_std):  # Sampling has been done using metropolis algorithm. Here sample is a vector X having 6 components
    samples = []
    current = initial # initial sample which we choose by ourselves. We equated initial sample to current sample
    num_accept = 0

    for _ in range(num_samples):
        while True:
            # Propose a new candidate from a normal distribution. Here proposal_std is the standard deviation which we choose by ourselves.
            candidate = np.random.normal(current, proposal_std)

            # Ensure the candidate falls within the specified bounds
            if (0 <= candidate[0] <= 5 and 0 <= candidate[3] <= 5 and
                0 <= candidate[1] <= np.pi and 0 <= candidate[4] <= np.pi and
                0 <= candidate[2] <= 2*np.pi and 0 <= candidate[5] <= 2*np.pi):

                candidate_value_f = f(candidate)
                candidate_value_p = p(candidate)

                # Discard if candidate value is NaN or inf
                if candidate_value_f != -np.inf or candidate_value_p != -np.inf:
                    break

        # Calculate acceptance probability
        acceptance_prob = min(1, candidate_value_p / p(current))

        # Accept or reject the candidate
        if np.random.uniform() < acceptance_prob:
            current = candidate
            num_accept += 1

        samples.append(current)

    return np.array(samples), num_accept

# Monte Carlo integration
def monte_carlo_integration(samples, f, p):

    values_f = np.array([f(sample) for sample in samples])
    # Filter out -inf values
    values_f = values_f[~np.isnan(values_f) & ~np.isinf(values_f)]

    values_p = np.array([p(sample) for sample in samples])
    # Filter out -inf values
    values_p = values_p[~np.isnan(values_p) & ~np.isinf(values_p)]

    values = values_f / values_p
    values = values[~np.isnan(values) & ~np.isinf(values)]

    return np.mean(values)

In [18]:
X_features[11021]

array([0.15015015, 0.09434212, 0.18868424, 0.05505506, 0.03459211,
       0.06918422])

In [19]:
# Parameters
initial_point = X_features[11021]
num_samples = 100000
proposal_std = 0.4

# Run Metropolis sampling
result = metropolis_sampling(f,p,initial_point, num_samples, proposal_std)

samples_f = result[0]
num_accept_f = result[1]

# Thermalization
burn_in = 1000
samples_f = samples_f[burn_in:]

# Perform Monte Carlo integration
integral_estimate_f = monte_carlo_integration(samples_f, f, p)
print(f"Estimated integral: {integral_estimate_f}")

Estimated integral: -0.15238579051704418


In [20]:
print("Number of Samples Collected: %s"%len(samples_f))
print("Number of Samples Accepted: %s"%(num_accept_f))
print("Fraction Acceptances: %s"%(num_accept_f / num_samples))

Number of Samples Collected: 99000
Number of Samples Accepted: 40491
Fraction Acceptances: 0.40491


In [21]:
# Parameters
initial_point = X_features[11021]
num_samples = 100000
proposal_std = 0.4

# Run Metropolis sampling
result = metropolis_sampling(g, p, initial_point, num_samples, proposal_std)

samples_g = result[0]
num_accept_g = result[1]

# Thermalization
burn_in = 1000
samples_g = samples_g[burn_in:]

# Perform Monte Carlo integration

integral_estimate_g = monte_carlo_integration(samples_g, g, p)
print(f"Estimated integral: {integral_estimate_g}")

Estimated integral: 0.012564902279554344


In [22]:
print("Number of Samples Collected: %s"%len(samples_g))
print("Number of Samples Accepted: %s"%(num_accept_g))
print("Fraction Acceptances: %s"%(num_accept_g / num_samples))

Number of Samples Collected: 99000
Number of Samples Accepted: 40487
Fraction Acceptances: 0.40487


In [23]:
E = integral_estimate_f / integral_estimate_g
print(f"E: {E}")

E: -12.127893009164655


1st run : -12.820260959503226  \\
2nd run : -12.553623879024876 \\
3rd run : -12.066661628712888 \\
4th run : -13.84645165511001 \\
5th run : -11.850889178794308 \\
6th run : -12.127893009164655